source/shared/cpp/ObjectModel/MarkDownParser.cpp

```diff
@@ -0,0 +1,430 @@
+#include <vector>
+#include <iomanip>
+#include <iostream>
+#include "MarkDownParser.h"
+
+using namespace AdaptiveCards;
+
+MarkDownParser::MarkDownParser (const std::string &txt) : m_currentDelimiterType (Init), 
+    m_currentWordIndex(0), m_lookBehind(Init), m_delimiterCnts(0), m_EmphasisState (EmphasisNotDetected), 
+    m_checkLookAhead(false), m_checkIntraWord(false), m_isRightEmphasisDetecting(false), m_text(txt), 
+    m_textBegin(m_text.begin()), m_curPos(m_text.begin()), m_textEnd(m_text.end()), m_LeftEmphasisDetecting(0),
+    m_leftLookUpTable (), m_rightLookUpTable (), m_tokenizedString(1, "")
+{
+}
+
+int MarkDownParser::adjustDelimsCntsAndMetaData(int leftOver, std::list<Emphasis>::iterator left_itr, 
+        std::list<Emphasis>::iterator right_itr)
+{
+    int delimCnts = 0;
+    if (leftOver >= 0)
+    {
+        delimCnts = left_itr->m_emphCnts - leftOver;
+        left_itr->m_emphCnts = leftOver;
+        right_itr->m_emphCnts = 0;
+    }
+    else
+    {
+        delimCnts = left_itr->m_emphCnts;
+        right_itr->m_emphCnts = leftOver * (-1);
+        left_itr->m_emphCnts = 0;
+    }
+    return delimCnts;
+}
+
+std::string MarkDownParser::TransformToHtml(void)
+{
+    GenSymbolTable ();
+    if (!m_leftLookUpTable.size () || !m_rightLookUpTable.size ())
+    {
+        return "<p>" + m_text + "</p>";
+    }
+
+    std::vector<std::list<Emphasis>::iterator> stk;
+    auto top_left = m_leftLookUpTable.begin(), top_right = m_rightLookUpTable.begin();
+    auto curr_left_delim = top_left, right_token_idx = top_right;
+    int lookBehind = (m_tokenizedString[top_left->m_idx][0] == '*')? Asterisk : Underscore; 
+    stk.push_back(top_left++);
+
+    while (!stk.empty() && top_right != m_rightLookUpTable.end())
+    {
+        // check if left most emphasis passed a new right emphasis delimiter
+        if (top_left != m_leftLookUpTable.end() && top_left->m_idx < top_right->m_idx)
+        {
+            stk.push_back(top_left);
+            lookBehind |= (m_tokenizedString[top_left->m_idx][0] == '*')? Asterisk : Underscore; 
+            ++top_left;
+        }
+        else
+        {
+            curr_left_delim = stk.back();
+            // because of rule #9 & #10 and multiple of 3 rule, left delim jump ahead of right delim,
+            // so need to check this condition.
```

Comment:
There should be a comment at some point pointing to the rules you are referring to.

Reply:
added the additional comment at line 28-29

---
In reply to: [154207523](https://github.com/Microsoft/AdaptiveCards/pull/1003#discussion_r154207523) [](ancestors = 154207523)